File: python/08_021.py
def query_rotated_array(arr, queries):
   """
   Compute the result of queries on a circular array 
   after rotations.
   
   Args:
       arr: List of integers representing the circular array
       queries: List of [rotation, index] pairs for each query
       
   Returns:
       List of integers representing the query results
   """
   n = len(arr)
   results = []
   
   for rotation, index in queries:
      # Calculate the effective rotation (in case rotation > n)
      effective_rotation = rotation % n
      
      # The element at position 'index' after rotating right by
      # 'effective_rotation' is originally at position
      # (index - effective_rotation) % n
      original_position = (index - effective_rotation) % n
      results.append(arr[original_position])
   
   return results

# Test with the example
arr = [3, 7, 1, 9, 5]
queries = [[2, 1], [4, 3], [0, 2]]
print(query_rotated_array(arr, queries))  # Output: [3, 3, 1]

File: java/08_021.java
import java.util.ArrayList;
import java.util.List;

/**
 * Efficiently query elements from a rotated circular array
 */
public class CircularArrayRotation {
   
   /**
    * Query elements after array rotations
    */
   public static List<Integer> queryRotatedArray(int[] arr, 
         int[][] queries) {
      int n = arr.length;
      List<Integer> results = new ArrayList<>();
      
      for (int[] query : queries) {
         int rotation = query[0];
         int index = query[1];
         
         // Calculate effective rotation (handle large values)
         int effectiveRotation = rotation % n;
         
         // Find original position before rotation
         int originalPosition = (index - effectiveRotation + n) % n;
         results.add(arr[originalPosition]);
      }
      
      return results;
   }
   
   public static void main(String[] args) {
      int[] arr = {3, 7, 1, 9, 5};
      int[][] queries = {{2, 1}, {4, 3}, {0, 2}};
      
      List<Integer> results = queryRotatedArray(arr, queries);
      System.out.println(results); // Output: [3, 3, 1]
   }
}